tensorboard/webapp/widgets/line_chart_v2/lib/formatter.ts

```diff
@@ -0,0 +1,122 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {format, scaleTime} from '../../../third_party/d3';
+
+export interface Formatter {
+  formatTick(x: number): string;
+  formatShort(x: number): string;
+  formatLong(x: number): string;
+}
+
+const d3NumberFormatter = format('.2~e');
+const d3TrimFormatter = format('~');
+const d3FloatFormatter = format('.4~');
+const d3LongFormatter = format(',~');
+
+function formatNumberShort(x: number): string {
+  if (x === 0) {
+    return '0';
+  }
+
+  const absNum = Math.abs(x);
+  if (absNum >= 100000 || absNum < 0.001) {
+    return d3NumberFormatter(x);
+  }
+
+  return d3TrimFormatter(x);
+}
+
+export const numberFormatter: Formatter = {
+  formatTick: formatNumberShort,
+  formatShort: formatNumberShort,
+  formatLong(x: number): string {
+    return d3LongFormatter(x);
+  },
+};
+
+const SECOND_IN_MS = 1000;
+const MINUTE_IN_MS = 60 * SECOND_IN_MS;
+const HOUR_IN_MS = 60 * MINUTE_IN_MS;
+const DAY_IN_MS = 24 * HOUR_IN_MS;
+const YEAR_IN_MS = 365 * DAY_IN_MS;
+
+function formatTime(x: number): string {
+  if (x === 0) return '0';
+
+  const sign = x <= 0 ? -1 : 1;
```

Comment:
This can just be `Math.sign(x)` (equivalent since you return early on
zeros anyway). But since your call site needs a conditional, anyway,
maybe easier to inline?

```javascript
  return `${x < 0 ? '-' : ''}${builder.join('')}`;
```

Seems more direct; fewer things to keep track of.